doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+  `BIO_write` or, if supported by the BIO, `BIO_sendmmsg`.
+
+A BIO may not necessarily be able to provide a poll descriptor. For example,
+memory-based BIOs such as `BIO_s_dgram_pair` do not correspond to any OS
+synchronisation resource, and thus the `BIO_get_rpoll_descriptor` and
+`BIO_get_wpoll_descriptor` calls are not supported for such BIOs.
+
+A BIO which supports these functions is known as pollable, and a BIO which does
+not is known as non-pollable. `BIO_s_dgram` supports these functions.
+
+The implementation of these functions for a `BIO_f_ssl` forwards to
+`SSL_get_rpoll_descriptor` and `SSL_get_wpoll_descriptor` respectively. The
+
+#### `BIO_s_dgram_mem`
+
+This is a basic memory buffer BIO with datagram semantics. Unlike
+`BIO_s_dgram_pair`, it is unidirectional and does not support peer addressing or
+local addressing.
+
+#### `BIO_err_is_non_fatal`
+
+A new predicate function `BIO_err_is_non_fatal` is defined which determines if
+an error code represents a non-fatal or transient error. For details, see
+[BIO_sendmmsg(3)](https://www.openssl.org/docs/manmaster/man3/BIO_sendmmsg.html).
+
+Q & A
+-----
+
+To assist in understanding, when a “TBD” listed above is removed, or when a
+relevant question is raised, the resolution to the question will be placed here.
+
+**Q. Should `SSL_do_handshake` wait until the handshake is completed, or until it
+is confirmed?**
+
+**Note:** [The terms *handshake complete* and *handshake confirmed* are defined
+in RFC 9001 and have specific
+meanings.](https://www.rfc-editor.org/rfc/rfc9001.html#name-handshake-complete)
+
+A. `SSL_do_handshake` should wait until the handshake is completed, because
+handshake completion represents the completion of the cryptographic
+authentication of the connection. When a connection's handshake is completed,
+TLS 1.3 Finished messages have been exchanged by both parties, even if the
+handshake has not yet been *confirmed*. Moreover, RFC 9001 s. 4.1.2 states:
+
+>Additionally, a client MAY consider the handshake to be confirmed when it
+>receives an acknowledgment for a 1-RTT packet.
+
+This logically implies that it is OK for a client to start transmitting 1-RTT
+packets prior to handshake confirmation, otherwise there would be no in-flight
+1-RTT packets for the client to receive ACKs for.
+
+**Q. Does `ENABLE_PARTIAL_WRITE` interact with blocking mode?**
+
+A. No; this mode is only relevant to non-blocking mode. In blocking mode,
+`SSL_write` always waits until all data is written unless an error occurs. The
+semantics of `SSL_write` are preserved unchanged.
+
+**Q. Does `SSL_write` block until data is written to the network, or simply
+until it is buffered?**
+
+A. `SSL_write` blocks until it has accepted responsibility for the data passed
+to it, just like `write(2)` or `send(2)`. In other words, it blocks until it can
+buffer the data. This does not necessarily mean that the data has actually been
```

Comment:
While this seems to be the right answer IMO it's unclear to me how this will interact with the IO solution. Hopefully that is covered in your other document.

Reply:
It's fairly simple, it goes into a `QUIC_SSTREAM` internally. So as not to buffer infinitely, if the `QUIC_SSTREAM` is full, we tick/block until there is more room in the `QUIC_SSTREAM` or the connection terminates, etc.